radix-engine-interface/src/math/decimal.rs

```diff
@@ -197,32 +198,63 @@ impl Decimal {
         }
     }
 
-    /// The square root of a Decimal. Uses a standard Babylonian method.
+    /// Square root of a Decimal
     pub fn sqrt(&self) -> Option<Self> {
         if self.is_negative() {
             return None;
         }
         if self.is_zero() {
             return Some(Self::ZERO);
         }
-        // Start with an arbitrary number as the first guess
-        let mut result = *self / Self::from(2u8);
-        // Too small to represent, so we start with self
-        // Future iterations could actually avoid using a decimal altogether and use a buffered
-        // vector, only combining back into a decimal on return
-        if result.is_zero() {
-            result = *self;
+
+        // The I256 i associated to a Decimal d is : i = d*10^18.
+        // Therefore, taking sqrt yields sqrt(i) = sqrt(d)*10^9 => We lost precision
+        // To get the right precision, we compute : sqrt(i*10^18) = sqrt(d)*10^18
+        let self_512: I512 = I512::from(self.0);
+        let correct_nb = self_512 * I512::from(Decimal::one().0);
```

Comment:
We might overflow by multiplying `self_512 * ONE`. To achieve the precision, we will have to promote it to `I1024`.